gym_madras.py

```diff
@@ -0,0 +1,160 @@
+        self.port = port
+        self.CLIENT_MAX_STEPS = CLIENT_MAX_STEPS
+        self.client_type = 0  # Snakeoil client type
+        self.initial_reset = True
+        self.early_stop = True
+        self.pid_assist = pid_assist
+        if self.pid_assist:
+            self.PID_latency = 10
+        else:
+            self.PID_latency = 1
+        self.accel_PID = PID(np.array([10.5, 0.05, 2.8]))  # accel PID
+        self.steer_PID = PID(np.array([5.1, 0.001, 0.000001]))  # steer PID
+
+        self.prev_lane = 0
+        self.prev_angle = 0
+        self.prev_vel = 0
+        self.prev_dist = 0
+        self.ob = None
+
+    def reset(self, prev_step_info=None):
+        """Reset Method."""
+        if self.initial_reset:
+            while self.ob is None:
+                try:
+                    self.client = snakeoil3.Client(p=self.port,
+                                                   vision=self.vision)
+                    # Open new UDP in vtorcs
+                    self.client.MAX_STEPS = self.CLIENT_MAX_STEPS
+                    self.client.get_servers_input(step=0)
+                    # Get the initial input from torcs
+                    raw_ob = self.client.S.d
+                    # Get the current full-observation
+                    self.ob = self.make_observation(raw_ob)
+                except:
+                    pass
+            self.initial_reset = False
+
+        else:
+            try:
+                if 'termination_cause' in list(prev_step_info.keys()) and\
+                        prev_step_info['termination_cause'] == 'hardReset':
+                    self.ob, self.client =\
+                        TorcsEnv.reset(self, client=self.client, relaunch=True)
+                else:
+                    self.ob, self.client =\
+                        TorcsEnv.reset(self, client=self.client, relaunch=True)
+
+            except Exception as e:
+                self.ob = None
+                while self.ob is None:
+                    try:
+                        self.client = snakeoil3.Client(p=self.port,
+                                                       vision=self.vision)
+                        # Open new UDP in vtorcs
+                        self.client.MAX_STEPS = self.CLIENT_MAX_STEPS
+                        self.client.get_servers_input(step=0)
+                        # Get the initial input from torcs
+                        raw_ob = self.client.S.d
+                        # Get the current full-observation from torcs
+                        self.ob = self.make_observation(raw_ob)
+                    except:
+                        pass
+
+        self.distance_traversed = 0
+        s_t = np.hstack((self.ob.angle, self.ob.track, self.ob.trackPos,
+                        self.ob.speedX, self.ob.speedY, self.ob.speedZ,
+                        self.ob.wheelSpinVel / 100.0, self.ob.rpm))
+
+        return s_t
+
+    def step(self, desire):
+        """Step Function."""
```

Comment:
Not a very useful comment. Please remove it or write it more elaborately. Eg: this function is executed for every time step etc etc.